### PR TITLE
Fixes/pcap timing/v18

### PR DESCRIPTION
**Pull request by victorjulien**
Fixes concurrency issues in 'autofp' mode while processing pcaps. See https://github.com/OISF/suricata/commit/9f1922e1756017b66c7161d49202a4120dd7f231 and https://github.com/OISF/suricata/commit/1e9333288f06ddc2f469053205a5d6662bf32ed7 for further explanations.

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):
Passed.

### CODE REVIEW

File: src/source-pcap-file-helper.c [hunk 1]
```diff
@@ -120,6 +120,14 @@ TmEcode PcapFileDispatch(PcapFileFileVars *ptv)
 {
     SCEnter();
 
+    /* initialize all the threads initial timestamp */
```
Comment (jlucovsky) [inline src/source-pcap-file-helper.c]: nit: `thread's`

File: src/source-pcap-file-helper.c [hunk 2]
```diff
@@ -160,6 +168,23 @@ TmEcode PcapFileDispatch(PcapFileFileVars *ptv)
     SCReturnInt(loop_result);
 }
 
+/** \internal
+ *  \brief get the timestamp of the first packet and rewind
+ *  \retval bool true on success, false on error
```
Comment (jlucovsky) [inline src/source-pcap-file-helper.c]: add param `pfv` to comments?

File: src/source-pcap-file-helper.c [hunk 3]
```diff
@@ -160,6 +168,23 @@ TmEcode PcapFileDispatch(PcapFileFileVars *ptv)
...
+ */
+static bool PeekFirstPacketTimestamp(PcapFileFileVars *pfv)
+{
+    int r = pcap_next_ex(pfv->pcap_handle, &pfv->first_pkt_hdr, &pfv->first_pkt_data);
+    if (r <= 0 || pfv->first_pkt_hdr == NULL) {
+        SCLogError(SC_ERR_PCAP_OPEN_OFFLINE,
+                "failed to get first packet timestamp. pcap_next_ex(): %d", r);
+        return false;
+    }
+    pfv->first_pkt_ts.tv_sec = pfv->first_pkt_hdr->ts.tv_sec;
```
Comment (jlucovsky) [inline src/source-pcap-file-helper.c]: Suggest  `pfv->first_pkt_ts = pfv->first_pkt_hdr->ts`

File: src/source-pcap-file-helper.c [hunk 4]
```diff
@@ -196,39 +221,40 @@ TmEcode InitPcapFile(PcapFileFileVars *pfv)
...
+    if (!PeekFirstPacketTimestamp(pfv))
+        SCReturnInt(TM_ECODE_FAILED);
+
     DecoderFunc temp;
     TmEcode validated = ValidateLinkType(pfv->datalink, &temp);
     SCReturnInt(validated);
 }
 
-TmEcode ValidateLinkType(int datalink, DecoderFunc *decoder)
+TmEcode ValidateLinkType(int datalink, DecoderFunc *DecoderFn)
```
Comment (jlucovsky) [inline src/source-pcap-file-helper.c]: style? parameter names are lowercase?

File: src/source-pcap-file-helper.h [hunk 5]
```diff
@@ -1,4 +1,4 @@
-/* Copyright (C) 2007-2010 Open Information Security Foundation
+/* Copyright (C) 2007-2020 Open Information Security Foundation
```
Comment (jlucovsky) [inline src/source-pcap-file-helper.h]: Copyright update for each module touched.

File: src/source-pcap-file-helper.h [hunk 6]
```diff
@@ -74,6 +74,12 @@ typedef struct PcapFileFileVars_
     struct bpf_program filter;
 
     PcapFileSharedVars *shared;
+
+    /* fields used to get the first packets timestamp early,
```
Comment (jlucovsky) [inline src/source-pcap-file-helper.h]: nit: `packet's`

File: src/tm-threads.c [hunk 7]
```diff
@@ -2213,6 +2216,7 @@ void TmThreadsUnregisterThread(const int id)
     SCMutexUnlock(&thread_store_lock);
 }
 
+#define COPY_TIMESTAMP(src,dst) ((dst)->tv_sec = (src)->tv_sec, (dst)->tv_usec = (src)->tv_usec) // XXX unify with flow-util.h
```
Comment (jlucovsky) [inline src/tm-threads.c]: Why not a simple structure copy?

File: src/tm-threads.c [hunk 8]
```diff
@@ -2223,33 +2227,72 @@ void TmThreadsSetThreadTimestamp(const int id, const struct timeval *ts)
...
+            break;
+        if (!(timercmp(&t->pktts, &nullts, ==))) {
+            /* ignore sleeping threads */
+            if (t->sys_sec_stamp + 1 < (uint32_t)systs.tv_sec)
+                continue;
+
             if (!set) {
-                local.tv_sec = t->ts.tv_sec;
-                local.tv_usec = t->ts.tv_usec;
+                local.tv_sec = t->pktts.tv_sec;
```
Comment (jlucovsky) [inline src/tm-threads.c]: structure copy?

File: src/source-pcap-file-helper.c (hunk 3 above)
Comment (victorjulien) [reply src/source-pcap-file-helper.c]: These are not guaranteed to be the same. More specifically, on OpenBSD they are not.

File: src/source-pcap-file-helper.c (hunk 4 above)
Comment (victorjulien) [reply src/source-pcap-file-helper.c]: Func pointers should be in this notation.

File: src/tm-threads.c (hunk 8 above)
Comment (victorjulien) [reply src/tm-threads.c]: ack

File: src/tm-threads.c (hunk 7 above)
Comment (victorjulien) [reply src/tm-threads.c]: Historical reasons. Changing this.

**Comment by victorjulien** [comment]
Merged through #4626. Will do another PR with some additional fixes/improvements.

**Pull request closed without merging**